### DISCUSSION TITLE
Authentication API and/or Access-Control-Allow-Credentials flag for a separate front-end

### DISCUSSION THREAD

**Discussion by ghost** [ckan/ckan#6022] [Ideas]
I am writing a separate JS front-end for CKAN and looking for the ways to authenticate into CKAN from the front-end. In version 2.9 came the feature of [API tokens](https://github.com/ckan/ckan/pull/5146), but still in order to get a token, one needs to log in somehow. I see here three possible solutions listed from the easiest to the most complex:

1) A user logs into JS front-end, which via HTTP POST sends the login+password to the CKAN back-end and saves CKAN's session cookies into user's browser.
For this to work one needs to: Add the front-end host to `ckan.cors.origin_whitelist`; Set `Access-Control-Allow-Credentials` to true in order to CKAN cookies to be saved when accessed via a front-end.

2) A user logs into CKAN by typing login+password after a successful login CKAN redirects the user to front-end along with API token.
For this to work one needs to: Modify CKAN login template page which redirects the user and provides API token in URL parameters; implement a logic in JS front-end for API token rotation as we probably don't want to use the same token over and over again on each login, probably some kind of generate/revoke token cycle is needed.

3) A user logs into JS front-end, which via a newly created API call sends the login+password to CKAN which responds with API token
For this to work one needs to: Create this new authentication API :-) and again implement some kind of token rotation unless it is implemented in API itself.

I am strongly leaning towards the first solution which requires the least amount of logic in JS front-end and the least amount of modifications into CKAN core. The change needed is to add a new flag say `who.allowcookies` which sets `Access-Control-Allow-Credentials` into [L96](https://github.com/ckan/ckan/blob/master/ckan/views/__init__.py#L96). Additionally, there could be a logic which also sets `who.samesite = Strict` automatically. This will mitigate CSRF attack risks.

Please, comment below if there are flaws in this reasoning.

**Comment by amercader** [comment]
@markpilkku We are moving away from the ideas repo to Discussions on this repo so I moved your comment here.

I know that the folks at Datopian have worked on this front and might have some insights to share here. cc @anuveyatsu @rufuspollock 

Also some other people has worked on detached frontend at some point, so feel free to share your experiences.

**Comment by anuveyatsu** [comment]
@markpilkku yes, as @amercader has mentioned we did some similar work. Here is the CKAN extension that provides API for login into CKAN: https://github.com/datopian/ckanext-auth

Hope this helps!
